JOSS paper (#265)
gyselax/gyselalibxx: docs/joss_paper/paper.md

- given-names: Philipp
  surname: Krah
  affiliation: 2
  orcid: 0000-0001-8982-4230

- given-names: Etienne
  surname: Malaboeuf
  affiliation: 5
  orcid: 0009-0007-3320-1406

- given-names: Kevin
  surname: Obrejan
  affiliation: 2
  orcid: 0000-0002-1906-4181

- given-names: Thomas
  surname: Padioleau
  affiliation: 3
  orcid: 0000-0001-5496-0013

- given-names: Matthieu
  surname: Protais
  affiliation: 2
  orcid: 0009-0003-5852-7446

- given-names: Pauline
  surname: Vidal
  affiliation: 4
  orcid: 0009-0008-7233-6351

affiliations:

- name: SCITAS, EPFL, CH-1015 Lausanne, Switzerland
  index: 1
- name: CEA, IRFM, 13108 Saint-Paul-lez-Durance Cedex, France
  index: 2
- name: >-
    Université Paris-Saclay, UVSQ, CNRS, CEA,
    Maison de la Simulation,
    91191, Gif-sur-Yvette, France
  index: 3
- name: >-
    Max-Planck-Institut für Plasmaphysik, Garching,
    Germany
  index: 4
- name: CINES, France
  index: 5

date: 24 April 2025
bibliography: paper.bib

---

## Summary

Gyselalib++ provides the mathematical building blocks to construct gyrokinetic plasma simulations in C++, simulating a distribution function discretised in phase space on a fixed grid.

> It relies on the DDC library[@ddc] to statically type the discretisation dimensions; thus preventing many common sources of errors.

Reviewer: I find it strange to read "...library(Padioleau ...)", without spaces between the word and the reference thereafter. Maybe include spaces in between?

Author: Fixed dccbc340e

Via DDC, Gyselalib++ also leverages the Kokkos framework [@trott2022], ensuring portability across both CPU and various GPU architectures.
The library provides a variety of tools including semi-Lagrangian advection operators, solvers for partial differential equations (PDEs), quadrature rules, and time steppers.
All operators are designed to function correctly or raise compiler errors on non-orthonormal coordinate systems thanks to the static typing.

## Statement of Need

Plasma simulations are essential for the development of magnetic confinement fusion devices for energy production.
The low collisionality of such plasmas make kinetic models a judicious choice.
In particular gyrokinetic theory [@brizard2007; @krommes2012], which reduces the 6D problem to a 5D problem by removing high frequency dynamics, is a popular tool for such simulations [@garbet2010].
Despite the reduction in dimensionality such simulations still require massively powerful high-performance computing (HPC) resources.

> For ITER-sized simulations, exa-scale resources would still be required.

Reviewer: _exascale_  without - ?

The pre-existing Gysela code [@grandgirard2016], written in Fortran, originally aimed to simulate plasma in the core region of a tokamak using semi-Lagrangian advection with a distribution function discretised in phase space on a fixed grid.

> This approach was shown to work well and allowed the study of many interesting physical phenomena [@Esteve2018;@Sarazin2021;@DifPradalier2022].

Reviewer: references do not show up in PDF ? ? ?

Author: Seems ok locally:
![image](https://github.com/user-attachments/assets/58afc60b-fbcf-46c9-b479-075be3b5cff8)

Reviewer: Maybe whitespaces missing between refs?

In Gyselalib++, equations are either expressed in tensor notation, so that they are either accurate for all geometries or do not compile, or they explicitly state their dependencies.
C++ further enables us to add static assertions for cases with restricted applicability to prevent their misuse.

In contrast to Gysela, Gyselalib++ has been conceived as a library, similar to the SeLaLib Fortran library [@selalib], whose independent elements are each unit tested and can be combined to build a final simulation.
This design makes the library more versatile, since its elements are not tied to a specific simulation and can be adapted to different needs.
The shared elements also provide more confidence in the reliability of the implementation as they can prove their validity across multiple applications.

Gyselalib++ includes a range of reusable mathematical operators for plasma simulations.
These include, but are not limited to, semi-Lagrangian advection schemes, numerical quadrature, differential operators (e.g. finite difference methods), and solvers for common partial differential equations.

> A complete list of available operators can be found in the [documentation](https://gyselax.github.io/gyselalibxx/).

Author: Done: [4162a9c](https://github.com/gyselax/gyselalibxx/pull/265/commits/4162a9c8283e720db0461eeeacba90a75ca8000f)

- given-names: Alexander
  surname: Hoffmann
  affiliation: 4
  orcid: 0009-0005-0736-6122

- given-names: Abdelhadi
  surname: Kara
  affiliation: 2
  orcid: 0009-0000-9693-5828

- given-names: Philipp
  surname: Krah
  affiliation: 2
  orcid: 0000-0001-8982-4230

- given-names: Etienne
  surname: Malaboeuf
  affiliation: 5
  orcid: 0009-0007-3320-1406

- given-names: Kevin
  surname: Obrejan
  affiliation: 2
  orcid: 0000-0002-1906-4181

- given-names: Thomas
  surname: Padioleau
  affiliation: 3
  orcid: 0000-0001-5496-0013

- given-names: Matthieu
  surname: Protais
  affiliation: 2
  orcid: 0009-0003-5852-7446

- given-names: Pauline
  surname: Vidal
  affiliation: 4
  orcid: 0009-0008-7233-6351

affiliations:

- name: SCITAS, EPFL, CH-1015 Lausanne, Switzerland
  index: 1
- name: CEA, IRFM, 13108 Saint-Paul-lez-Durance Cedex, France
  index: 2
- name: >-
    Université Paris-Saclay, UVSQ, CNRS, CEA,
    Maison de la Simulation,
    91191, Gif-sur-Yvette, France
  index: 3
- name: >-
    Max-Planck-Institut für Plasmaphysik, Garching,
    Germany
  index: 4
- name: CINES, France
  index: 5

date: 24 April 2025
bibliography: paper.bib

---

## Summary

> Gyselalib++ provides the mathematical building blocks to construct gyrokinetic plasma simulations in C++, simulating a distribution function discretised in phase space on a fixed grid.

Reviewer 2: If you think they are equivalent, I am fine

> Gyselalib++ provides the mathematical building blocks to construct gyrokinetic plasma simulations in C++, simulating a distribution function discretised in phase space on a fixed grid.

Author: ChatGPT agrees with you though :sweat_smile:

It relies on the DDC library [@ddc] to statically type the discretisation dimensions; thus preventing many common sources of errors.
Via DDC, Gyselalib++ also leverages the Kokkos framework [@trott2022], ensuring portability across both CPU and various GPU architectures.

> The library provides a variety of tools including semi-Lagrangian advection operators, solvers for partial differential equations (PDEs), quadrature rules, and time steppers.

Reviewer 2: ```suggestion
The library provides a variety of tools including semi-Lagrangian advection operators, solvers for partial differential equations (PDEs), quadrature rules, and time integrators.
```

All operators are designed to function correctly or raise compiler errors on non-orthonormal coordinate systems thanks to the static typing.

## Statement of Need

Plasma simulations are essential for the development of magnetic confinement fusion devices for energy production.
The low collisionality of such plasmas make kinetic models a judicious choice.
In particular gyrokinetic theory [@brizard2007; @krommes2012], which reduces the 6D problem to a 5D problem by removing high frequency dynamics, is a popular tool for such simulations [@garbet2010].
Despite the reduction in dimensionality such simulations still require massively powerful high-performance computing (HPC) resources.
For ITER-sized simulations, exascale resources would still be required.

> The pre-existing Gysela code [@grandgirard2016], written in Fortran, originally aimed to simulate plasma in the core region of a tokamak using semi-Lagrangian advection with a distribution function discretised in phase space on a fixed grid.

Reviewer 3: Is it Gysela? GYSELA? or GYSELA-X?

Reviewer 3: In https://www.sciencedirect.com/science/article/pii/S0010465516301230?via%3Dihub It seems it was "GYSELA".

Author: Fixed 94f7606fd

Author: Good catch

> The pre-existing GYSELA code [@grandgirard2016], written in Fortran, originally aimed to simulate plasma in the core region of a tokamak using semi-Lagrangian advection with a distribution function discretised in phase space on a fixed grid.

Author: I prefer the old wording. Within the Gysela team we often talk about the "GYSELA Fortran code" to differentiate it from the C++ code but here I think it is better to have the citation close to the name

This approach was shown to work well and allowed the study of many interesting physical phenomena [@Esteve2018;@Sarazin2021;@DifPradalier2022].
However expanding this code to use more complex mathematical methods such as non-uniform points (vital for handling the different magnitudes of physical quantities in the core and edge regions), and increasingly complex geometries (such as D-shape geometries, geometries including both closed and open field lines, and potentially stellarator geometries) has proved to be challenging and sometimes error-prone.
The challenges of such extensions are further amplified when trying to organise such a code for use on new GPU architectures, necessary for exascale simulations.
This is a challenge shared by other gyrokinetic codes [@trilaksono2023].

> In the case of Gysela, several of the required changes would have affected a large percentage of the code base.

Reviewer 3: same question here.

Author: Fixed 94f7606fd

> This means that the effort required was comparable to a complete rewrite, without the advantages such a rewrite can bring.

Reviewer 2: This sentence is not very clear to me. Can you rephrase this?

Author: Is this clearer: 17d773cfe ?

Reviewer 2: Better for me, thanks

For example, we have been able to capitalise on C++'s strengths by using template programming to enforce the correctness of the implemented equations.
A common source of error is writing equations with implicit assumptions, such as assuming an orthonormal coordinate system, or specific properties like those of a circular coordinate system.
In Gyselalib++, equations are either expressed in tensor notation, so that they are either accurate for all geometries or do not compile, or they explicitly state their dependencies.

> C++ further enables us to add static assertions for cases with restricted applicability to prevent their misuse.

Reviewer 2: Can you refer to DDC somewhere in order to explain how DDC helps to prevent many common sources of errors. This is explained in summary, but not explained in the body

Author: Do you have a suggestion of where?

> This is explained in summary, but not explained in the body

The summary is not equivalent to an abstract. It is just one of the sections of the paper.

Author: I can add something at the end of the 3rd paragraph. But if this needs doing for DDC then it probably also needs to be done for Kokkos. If I add both then I should chop something else to get back closer to the word limit. Do you see anything that you think is less important than repeating DDC?

Reviewer 2: > The summary is not equivalent to an abstract. It is just one of the sections of the paper.

OK

> Do you have a suggestion of where?

If you can explain more in summary, that is reasonable as well.
For readers that are not very familiar with DDC, the following phrase is difficult to understand I guess
> It relies on the DDC library [@ddc] to statically type the discretisation dimensions; thus preventing many common sources of errors.

> But if this needs doing for DDC then it probably also needs to be done for Kokkos.

I think Kokkos is famous enough, so we do not need to explain more.
If you say Kokkos ensures performance portability, it is good enough

Author: I have added this: c72e27b8c

Reviewer 2: Looks good

> In contrast to Gysela, Gyselalib++ has been conceived as a library, similar to the SeLaLib Fortran library [@selalib], whose independent elements are each unit tested and can be combined to build a final simulation.

Reviewer 2: ```suggestion
In contrast to Gysela, Gyselalib++ has been conceived as a library, similar to the SeLaLib Fortran library [@selalib], whose independent elements are each unit-tested and can be combined to build a final simulation.
```

Author: Fixed [02640b8](https://github.com/gyselax/gyselalibxx/pull/265/commits/02640b88ca2c6fac08aa24844bb8fc5c99a4c70f)

This design makes the library more versatile, since its elements are not tied to a specific simulation and can be adapted to different needs.
The shared elements also provide more confidence in the reliability of the implementation as they can prove their validity across multiple applications.

Gyselalib++ includes a range of reusable mathematical operators for plasma simulations.

> These include, but are not limited to, semi-Lagrangian advection schemes, numerical quadrature, differential operators (e.g. finite difference methods), and solvers for common partial differential equations.

Reviewer 3: ```suggestion
These include, but are not limited to, semi-Lagrangian advection schemes, numerical quadrature, differential operators (e.g. finite difference methods), and solvers for common PDEs.
```

(if you want to reduce 2 words knowning that you have already introduced the notation "PDE2s")

Author: Ah! I didn't realise I already had PDEs above. Thanks. Fixed 94f7606fd

While these examples are included primarily for illustration, they also serve as valuable test-beds for developing and validating new numerical methods.
As such, these examples are ideal for mathematicians looking to validate new methods in realistic, publication-ready test cases.

## Acknowledgements

This project has received funding from the EoCoE-III project (Energy-oriented Centre of Excellence for Exascale HPC applications).
This work has been carried out within the framework of the EUROfusion Consortium, funded by the European Union via the Euratom Research and Training Program (Grant Agreement No. 101052200 — EUROfusion).
Views and opinions expressed are those of the author(s) only and do not necessarily reflect those of the European Union or the European Commission.
Neither the European Union nor the European Commission can be held responsible for them.
Emily Bourne’s salary was paid for by the EUROfusion Advanced Computing Hub (ACH).

> This project was provided with computer and storage resources by GENCI CINES thanks to the grant 2024-A0160502224 on the supercomputer  on the supercomputer Adastra’s the GENOA partition.

Reviewer 4: Shall we also acknowledge the PTCs ?

> We acknowledge the financial support of the Cross-Disciplinary Program on "Numerical simulation" of CEA, the French Alternative Energies and Atomic Energy Commission.

Author: [1ce89e7](https://github.com/gyselax/gyselalibxx/pull/265/commits/1ce89e78621f1b66ed1e347c1ba89d9ed119f21b)

> ## References

Reviewer 3: I am not really aware of paper written in markdown, but this selection seems to be empty. All the references will be automatically listed here? You dont need to use an equivalent of "/printbibliographie" ?

Author: I think so. This is for the volume/issue/page in the Journal of Open Source Software. It's not published yet so it doesn't have these

Reviewer 3: Ok :+1:

  orcid: 0000-0002-0015-4304

- given-names: Peter
  surname: Donnel
  affiliation: 2
  orcid: 0000-0002-6669-416X

- given-names: Alexander
  surname: Hoffmann
  affiliation: 4
  orcid: 0009-0005-0736-6122

- given-names: Abdelhadi
  surname: Kara
  affiliation: 2
  orcid: 0009-0000-9693-5828

- given-names: Philipp
  surname: Krah
  affiliation: 2
  orcid: 0000-0001-8982-4230

- given-names: Etienne
  surname: Malaboeuf
  affiliation: 5
  orcid: 0009-0007-3320-1406

- given-names: Kevin
  surname: Obrejan
  affiliation: 2
  orcid: 0000-0002-1906-4181

- given-names: Thomas
  surname: Padioleau
  affiliation: 3
  orcid: 0000-0001-5496-0013

- given-names: Matthieu
  surname: Protais
  affiliation: 2
  orcid: 0009-0003-5852-7446

- given-names: Pauline
  surname: Vidal
  affiliation: 4
  orcid: 0009-0008-7233-6351

affiliations:

- name: SCITAS, EPFL, CH-1015 Lausanne, Switzerland
  index: 1
- name: CEA, IRFM, 13108 Saint-Paul-lez-Durance Cedex, France
  index: 2
- name: >-
    Université Paris-Saclay, UVSQ, CNRS, CEA,
    Maison de la Simulation,
    91191, Gif-sur-Yvette, France
  index: 3
- name: >-
    Max-Planck-Institut für Plasmaphysik, Garching,
    Germany
  index: 4

> - name: CINES, France

Author: Fixed 5583bc623bb3e5bf1b288064c605f62c511b55b6

  index: 5

date: 24 April 2025
bibliography: paper.bib

---

## Summary

Gyselalib++ provides the mathematical building blocks to construct gyrokinetic plasma simulation codes in C++, simulating a distribution function discretised in phase space on a fixed grid.
It relies on the DDC library [@ddc] to statically type the discretisation dimensions; thus preventing many common sources of errors.
Via DDC, Gyselalib++ also leverages the Kokkos framework [@trott2022], ensuring performance portability across various CPU and GPU architectures.

> The library provides a variety of tools including semi-Lagrangian advection operators, solvers for partial differential equations (PDEs), quadrature rules, and time integrators.

Reviewer 5: Hmm.. Maybe what you want to use "elliptic and hyperbolic PDEs" i.e. PDEs that are, or not, time-dependent

> All operators are designed to function correctly or raise compiler errors on non-orthonormal coordinate systems thanks to the static typing.

Reviewer 6: Ok since it's the summary section, I assumed it was a rundown of what's said later in more details. That new sentence you propose is much clearer.

> All operators are designed to function correctly or raise compiler errors on non-orthonormal coordinate systems thanks to the static typing.

Author: Done : 4227e9fe4

## Statement of Need

Plasma simulations are essential for the development of magnetic confinement fusion devices for energy production.

> The low collisionality of such plasmas make kinetic models a judicious choice.

Author: Done 275a56311e50fb52d4e7a9ab03d75988ed48cc66

In particular gyrokinetic theory [@brizard2007; @krommes2012], which reduces the 6D problem to a 5D problem by removing high frequency dynamics, is a popular tool for such simulations [@garbet2010].
Despite the reduction in dimensionality such simulations still require massively powerful high-performance computing (HPC) resources.
For ITER-sized simulations, exascale resources would still be required.

> The pre-existing GYSELA code [@grandgirard2016], written in Fortran, originally aimed to simulate plasma in the core region of a tokamak using semi-Lagrangian advection with a distribution function discretised in phase space on a fixed grid.

Author: Fixed: 78217ff24

This approach was shown to work well and allowed the study of many interesting physical phenomena [@Esteve2018;@Sarazin2021;@DifPradalier2022].

> However expanding this code to use more complex mathematical methods such as non-uniform points (vital for handling the different magnitudes of physical quantities in the core and edge regions), and increasingly complex geometries (such as D-shape geometries, geometries including both closed and open field lines, and potentially stellarator geometries) has proved to be challenging and sometimes error-prone.

Reviewer 6: "geometries including both open and closed field lines, X-points, ..." (open fields which exit the simulation box are a problem for flux-surface averaging, so it's not the same as the X-point's singularity)

> However expanding this code to use more complex mathematical methods such as non-uniform points (vital for handling the different magnitudes of physical quantities in the core and edge regions), and increasingly complex geometries (such as D-shape geometries, geometries including both closed and open field lines, and potentially stellarator geometries) has proved to be challenging and sometimes error-prone.

Author: Done : 896e1f02d

> However expanding this code to use more complex mathematical methods such as non-uniform points (vital for handling the different magnitudes of physical quantities in the core and edge regions), and increasingly complex geometries (such as D-shape geometries, geometries including both open and closed field lines, X-points, and potentially stellarator geometries) has proved to be challenging and sometimes error-prone.

Reviewer 5: Is non-uniformity used to handle different magnitudes of physical quantities or steep gradients of these quantities?

Reviewer 5: Ok, makes sense. I agree you can keep it like this

The challenges of such extensions are further amplified when trying to organise such a code for use on new GPU architectures, necessary for exascale simulations.

> This is a challenge shared by other gyrokinetic codes [@trilaksono2023].

Author: The middle challenge has been changed to complexities
275a56311e50fb52d4e7a9ab03d75988ed48cc66

---

## Summary

Gyselalib++ provides the mathematical building blocks to construct gyrokinetic plasma simulations in C++, simulating a distribution function discretised in phase space on a fixed grid.
It relies on the DDC library[@ddc] to statically type the discretisation dimensions; thus preventing many common sources of errors.
Via DDC, Gyselalib++ also leverages the Kokkos framework, ensuring portability across both CPU and various GPU architectures.
The library provides a variety of tools including semi-Lagrangian advection operators, solvers for partial differential equations (PDEs), quadrature rules, and time steppers.
All operators are designed to function correctly or raise compiler errors on non-orthonormal coordinate systems thanks to the static typing.

## Statement of Need

Plasma simulations are essential for the development of magnetic confinement fusion devices for energy production.
The low collisionality of such plasmas make kinetic models a judicious choice.
In particular gyrokinetic theory[@brizard2007; @krommes2012], which reduces the 6D problem to a 5D problem by removing high frequency dynamics, is a popular tool for such simlations [@garbet2010].

> Despite the reduction in dimensionality scuh simulations still require massively powerful high-performance computing (HPC) resources.

Author: ```suggestion
Despite the reduction in dimensionality such simulations still require massively powerful high-performance computing (HPC) resources.
```

This design makes the library more versatile, enabling users to rapidly assemble a wide range of simulations, including high-dimensional test cases.
The shared elements also provide more confidence in the reliability of the implementation, as they can prove their validity across multiple applications.

## State of the Field

Most established gyrokinetic simulations, such as GENE-X [@MICHELS2021107986] and GT5D [@Idomura_2009], are written in Fortran as stand-alone codes.
This  limits code sharing and reuse between projects.

> In contrast, many particle-in-cell codes are now developed around reusable libraries such as Smilei [@DEROUILLAT2018351] and AMReX [@AMReX_JOSS].

Reviewer 2: I will have a look

> In contrast, many particle-in-cell codes are now developed around reusable libraries such as Smilei [@DEROUILLAT2018351] and AMReX [@AMReX_JOSS].

Reviewer 2: XGC original is this one https://doi.org/10.1063/1.1707024
```bibtex
@article{10.1063/1.1707024,
    author = {Chang, C. S. and Ku, Seunghoe and Weitzner, H.},
    title = {Numerical study of neoclassical plasma pedestal in a tokamak geometry},
    journal = {Physics of Plasmas},
    volume = {11},
    number = {5},
    pages = {2649-2667},
    year = {2004},
    month = {05},
    abstract = {The fundamental properties of steep neoclassical plasma pedestals in a quiescent tokamak plasma have been investigated with a new guiding center particle code XGC: an X-point included Guiding Center code. It is shown that the width of the steepest neoclassical pedestals is similar to an experimentally observed edge pedestal width, and that a steep pedestal must be accompanied by a self-consistent negative radial electric field well. It is also shown that a steep neoclassical pedestal can form naturally at a quiescent diverted edge as the particle source from the neutral penetration (and heat flux from the core plasma) is balanced by the sharply increasing convective ion loss toward the separatrix. The steep neoclassical pedestal and the strong radial electric field well are suppressed by an anomalous diffusion coefficient of a strength appropriate to an L-mode state; nonetheless, the E×B shearing rate increases rapidly with pedestal temperature. Additionally, the present study shows that a steep pedestal at the diverted edge acts as a cocurrent parallel momentum source.},
    issn = {1070-664X},
    doi = {10.1063/1.1707024},
    url = {https://doi.org/10.1063/1.1707024},
    eprint = {https://pubs.aip.org/aip/pop/article-pdf/11/5/2649/19290044/2649\_1\_online.pdf},
}
```

Cabana is https://doi.org/10.21105/joss.04115
```bibtex
@article{Slattery_JOSS_2022,
  doi = {10.21105/joss.04115},
  url = {https://doi.org/10.21105/joss.04115},
  year = {2022},
  publisher = {The Open Journal},
  volume = {7},
  number = {72},
  pages = {4115},
  author = {Stuart Slattery and Samuel Temple Reeve and Christoph Junghans and Damien Lebrun-Grandié and Robert Bird and Guangye Chen and Shane Fogerty and Yuxing Qiu and Stephan Schulz and Aaron Scheinberg and Austin Isner and Kwitae Chong and Stan Moore and Timothy Germann and James Belak and Susan Mniszewski},
  title = {Cabana: A Performance Portable Library for Particle-Based Simulations},
  journal = {Journal of Open Source Software}
}
```

> In contrast, many particle-in-cell codes are now developed around reusable libraries such as Smilei [@DEROUILLAT2018351] and AMReX [@AMReX_JOSS].

Reviewer 2: For XGC, they recommend this rather than the original one. https://xgc.pppl.gov/html/general_info.html

```bibtex
@misc{ doecode_12570,
title = {XGC},
author = {Ku, Seung-Hoe and Hager, Robert and Scheinberg, Aaron and Dominski, Julien and Sharma, Amil and Churchill, Michael and Choi, Jong and Sturdevant, Ben and Mollén, Albert and Wilkie, George and Chang, Choong-Seock and Yoon, Eisung and Adams, Mark and Seo, Janghoon and Koh, Sehoon and D'Azevedo, Eduardo and Abbott, Steve and Worley, Patrick H. and Ethier, Stephane and Park, Gunyoung and Lang, Jianying and MacKie-Mason, Brian and Germaschewski, Kai and Suchyta, Eric and Carey, Varis and Cole, Michael and Trivedi, Pallavi and Chowdhury, Jugal},
abstractNote = {XGC (X-point included gyrokinetic code), Version 3 Primary Author: Seunghoe Ku (PPPL, sku@pppl.gov), Robert Hager (PPPL, rhager@pppl.gov) 31 July 2017 XGC is a gyrokinetic particle-in-cell code, which is sepeciallized in tokamak edge simulatin. The simulation domain can include the magnetic separatrix, magnetic axis and the biased material wall. XGC can run in full-f, total-delta-f, and conventional delta-f mode. The ion is kinetic always except ETG simulation. The electron can be adiabatic, fluid, drift kinetic, or gyrokinetic (for ETG). XGC is written in Fortran 90 and is designed for HPCs utilizing MPI, OpenMP, CUDA (GPU), OpenACC (GPU), and vectorization (Intel MIC-KNL). The weak scaling is roughly linear to the maximal nodes of leading HPCs in US. There several versions of XGCs for different purpose: XGC0 is earlist version of XGC0 which is designed for neoclassical transport in the tokamak edge. It uses full-f method, and 00-mode electrostatic field is solved only. RMP and current response can be calculated with coupling with M3D. XGC1 is for turbulence simulation with low parallel wavenumber. Piecewise field following coordinates are used to handle low k-parallel perturbation with small (~64) number of toroidal resolution. In total-delta-f method, it utilizes phase space grid in addition to particles, to handel non-maxwellian distribution in the tokamak edge. XGCa is axisymmetric version of XGC1 to simulate neoclassical transport. References: [1] S. Ku et al., Nuclear Fusion 49, 115021 (2009) [2] S. Ku, R. Hager, C.S. Chang et al., J. Comp. Physics, 315, 467 (2016) https://doi.org/10.1016/j.jcp.2016.03.062 [2] R. Hager. E.S. Yoon. S.Ku et al., J. Comp. Physics, 315, 644 (2016) https://doi.org/10.1016/j.jcp.2016.03.064 [3] R. Hager, J. Lang et al., Phys. Plasmas 24,054508 (2017) http://dx.doi.org/10.1063/1.4983320},
doi = {10.11578/dc.20180627.11},
url = {https://doi.org/10.11578/dc.20180627.11},
howpublished = {[Computer Software] \url{https://doi.org/10.11578/dc.20180627.11}},
year = {2018},
month = {jun}
}
```
